File: 0815_Bus Routes/BusRoutes.cs
public class Solution {
    public int NumBusesToDestination(int[][] routes, int source, int target) {
        if(routes == null || routes.Length == 0) return -1;
        if(source == target) return 0;
        var graph = new Dictionary<int, List<int>>();
        for(int i=0;i<routes.Length;i++){
            var stops = routes[i];
            foreach(var s in stops){                
                if(!graph.ContainsKey(s))    
                    graph.Add(s, new List<int>());
                graph[s].Add(i);
            }
        }
        // key is stop, value is route
        var dict = new Dictionary<int, int>();
        // check if the route processed or not
        var visited = new HashSet<int>();
        
        var queue = new Queue<int>();

        queue.Enqueue(source);
        var buses = 0;
        
        while(queue.Count > 0)
        {
            buses++;
            var size = queue.Count;
            
            while(size-- > 0)
            {
                var currStop = queue.Dequeue();
                
                foreach(var route in graph[currStop])
                {
                    if(visited.Contains(route)) continue;
                    visited.Add(route);
                    foreach(var stop in routes[route]){
                        if(stop == target) return buses;  
                        queue.Enqueue(stop);
                    }
                }
            }
        }
        
        return -1;
        
    }
}

File: 0815_Bus Routes/BusRoutes.java
class Solution {
    public int numBusesToDestination(int[][] routes, int S, int T) {
        if(routes == null || routes.length == 0) return 0;
        if(S == T) return 0;
        // build a graph, key is stop, value is the list of routes for the stop
        Map<Integer,List<Integer>> graph = new HashMap();
        for(int route = 0; route < routes.length; route++) {
            for(int stop : routes[route]) {
                graph.putIfAbsent(stop, new ArrayList<Integer>());
                graph.get(stop).add(route);
            }
        }
        
        int ans = 0;
        Queue<Integer> q = new LinkedList<Integer>();
        q.add(S);
        Set<Integer> visited = new HashSet();
        
        while(q.size() > 0) {
            int count = q.size();
            ans++;
            
            while(count-- > 0) {
                int currStop = q.poll();
                for(Integer route : graph.get(currStop)) {
                    if(visited.contains(route)) continue;
                    visited.add(route);
                    for(int nextStop : routes[route]) {
                        if(nextStop == T) return ans;
                        q.add(nextStop);
                    }
                }
            }
        }
        
        return -1;
    }
}